DHIS2-4857 - Support event layer downloads with multiple Feature Key formats.

amcgee opened the pull request:
This turned into a pretty big pull request - I will submit it for review as is and make modifications to the MetaData output format options in another change once this has been merged.  The major changes here are:

* Add `DataDownloadDialog` for downloading layer data
* Support customization of event layer GeoJSON key formats
* Show a loading spinner while downloading Event layer data at download-time (data is not available locally so it must be requested from the server)
* Show an error message in the dialog if the download fails for some reason
* Use a `redux-observable` epic to manage the download asynchronous behavior

**Tangential**

* Refactor `eventLoader.js` for less repetition and more reusability
* Move helper functions to `util/geojson.js` to keep `eventLoader`, `epics/dataDownload`, and `DataDownloadDialog`
* Implement unit tests with 100% coverage for `util/geojson.js`
* Add `Netlify.toml` to automatically deploy branches and PRs, like [this](https://deploy-preview-36--dhis2-maps.netlify.com/) for this PR.  Once merged this will be available on Master and on all PRs to Master
* Remove custom styling from Maps dialogs and implement DHIS2 identity-compliant Dialog styles in the dhis2 MUI theme so that they are automatically adopted in any MUI dialog.  **TODO**: move this to `d2-ui` and remove customization from the various dialogs in `d2-ui` components.

Dialog now looks like this for Event and Thematic layers, respectively (showing error state for reference):

<img width="496" alt="screenshot 2018-10-29 14 56 19" src="https://user-images.githubusercontent.com/947888/47732230-10b6f200-dc66-11e8-9ecc-709f6545149a.png">
<img width="493" alt="screenshot 2018-10-29 14 57 03" src="https://user-images.githubusercontent.com/947888/47732231-10b6f200-dc66-11e8-8e5c-84cee061738a.png">

**NB** Comment updated and WIP title removed, requesting review from @turban and @joeDHIS - you can test these changes [here](https://deploy-preview-36--dhis2-maps.netlify.com/) or check them out locally.

amcgee commented:
Hey @joeDHIS and @turban, I'd love some feedback on this Event Download implementation!

- I've implemented a dialog for all data downloads, including a link to [geojson.org](http://geojson.org) for some information on geojson (@turban should we link somewhere else?).
- For Event layers I've allowed the user to select the output format (ID, Code, or Name) for data dimensions and to select Human Readable (Name) or ID for all other GeoJSON keys.  This works fine, but it's a little clunky and inconsistent across different layer types.  I'm not sure how to get the ID/Code/Name from the webapi for other layer types (if it's even possible) and it would require a bit more refactoring to make that work in the code.

So three things I could use some help with:
1) Feedback on the dialog content, loading behavior, and output formats
2) For longer blocks of text like this, how do we usually do translation?  I think we probably don't want to inline the full text as the translation key?
3) Since the OutputIdScheme parameter appears to be broken in the backend I had to work around it here, feedback on the workaround would be great too. @janhenrikoverland @edoardo 

I'm also testing continuous deployment of branches and PRs, so **you can see a working version of this change** (pointing at the play/dev server) [here](https://deploy-preview-36--dhis2-maps.netlify.com/).  It will be automatically updated when the PR is changed, and you can also get to the link from the "netlify/dhis2-maps/deploy-preview" Check below.  NB: I had to manually add this url to the CORS whitelist but otherwise everything was automatic, yay!  👍 

Thanks!

amcgee commented:
Here's a screen recording for those who don't want to visit the [live demo](https://deploy-preview-36--dhis2-maps.netlify.com/)

![maps-36-demo](https://user-images.githubusercontent.com/947888/47425265-20c46280-d78a-11e8-90d0-5ee4bff8b300.gif)

amcgee commented:
And another to show the difference when downloading a thematic layer (in which case we already have the data locally, so no download delay, but also don't fetch updated key names so there's no option for the user there.

![maps-36-demo2](https://user-images.githubusercontent.com/947888/47425692-70575e00-d78b-11e8-90e4-97fbcd5e7679.gif)

turban commented:
Nice work! 

geojson.org is not very descriptive of the standard. I would suggest to remove the link and write something like: 

"The data for this layer will be downloaded in GeoJSON format. This format is supported my most mapping programs like QGIS and ArcGIS."

We could later link to a tutorial in our docs about how to import and use data in QGIS.

turban commented:
Small bug: "More actions" tooltip is displayed above the download dialog.  

![skjermbilde 2018-10-24 kl 14 59 16](https://user-images.githubusercontent.com/548708/47432028-80783900-d79d-11e8-8429-68736b64b725.png)

amcgee commented on src/loaders/eventLoader.js:
```diff
@@ -188,23 +163,73 @@ export const getAnalyticsRequest = async (
         (props, value, i) => ({
             ...props,
-            [headers[i].name]: names[value] || value,
+            [names[headers[i].name] || headers[i].name]: value,
```
@turban I believe the code was incorrect here before since names comes from the header list. I updated it to work for Event Download but would love your review to make sure I didn't change the behavior of this function.

turban commented:
Bug: Add event layer -> Program: Inpatient morbidity and mortality -> Add layer -> 

TypeError: Cannot read property 'outputIdScheme' of undefined

joeDHIS commented:
There are some guidelines that have been created for d2ui, which can be seen [here](https://invis.io/A7LT4TJYETS#/324707554_molecules). They haven't been built yet though. Looking at the examples, the "Add new Event Layer' looks correct.

Everything else looks nice. One question - do we want to inform the user why there is no option for setting key names for thematic layers? It's not immediately obvious to me, but I don't have any experience with GIS data, so maybe it's something our expected audience will understand. If not, a simple "Format for GeoJSON features keys cannot be modified for thematic layers", could suffice. That way a user isn't searching around, thinking they can't find the option or something. I'm happy to defer to you guys on this one and trust what you think.

turban commented:
Comparsion of the different GeoJSON-formats provided for event layer download:

1. Meta-data ID Format: ID
![id](https://user-images.githubusercontent.com/548708/47436046-04ceba00-d7a6-11e8-904e-acb07b6ce189.png)

2. Meta-data ID Format: Code
![code](https://user-images.githubusercontent.com/548708/47436067-0e582200-d7a6-11e8-88da-ff007981385d.png)

3. Meta-data ID Format: Name
![name](https://user-images.githubusercontent.com/548708/47436078-17e18a00-d7a6-11e8-8a2c-aee5a3a3020a.png)

4. Meta-data ID Format: Name + human-readable keys
![name_readable](https://user-images.githubusercontent.com/548708/47436091-1e700180-d7a6-11e8-8444-39b6f4b6b825.png)

I think 4 is the most readable and user friendly format - and I would suggest it to be the default.

@joeDHIS are we using Title Case or Sentence case for text field labels, titles etc.?

turban commented:
The text field is very long and the content is narrow - better to set a shorter max length than 100% or make the dialog smaller? 

![skjermbilde 2018-10-24 kl 16 13 53](https://user-images.githubusercontent.com/548708/47436977-e2d63700-d7a7-11e8-99c7-76cabfcbedf7.png)

@joeDHIS are we using Title Case or Sentence case for text field labels, titles etc.?

turban commented on src/loaders/eventLoader.js:
```diff
@@ -65,30 +49,16 @@ const eventLoader = async layerConfig => {
-                )
-            )
-            .filter(feature => isValidCoordinate(feature.geometry.coordinates));
+        const { names, data, response } = await loadData(analyticsRequest);
```
```suggestion
        const { names, data, response } = await loadData(analyticsRequest, config);
```

amcgee commented:
Good catch , yeah that is the change @turban thanks!

amcgee commented:
I matched the text casing in the Pivot Table app download menu

amcgee commented:
Feedback welcome and appreciated on styling, I haven't taken the time to fine-tuned it yet so there are inconsistencies.

amcgee commented:
> Small bug: "More actions" tooltip is displayed above the download dialog.
> 
> ![skjermbilde 2018-10-24 kl 14 59 16](https://user-images.githubusercontent.com/548708/47432028-80783900-d79d-11e8-8429-68736b64b725.png)

Yeah, it's an annoying bug but unrelated to this change.  Something to do with the combination of tooltip and popup menu, when the menu is dismissed the tooltip reappears until you enter/exit the hover state again.

amcgee commented:
Unit test coverage of `util/geojson.js` file:

<img width="706" alt="screenshot 2018-10-30 17 08 52" src="https://user-images.githubusercontent.com/947888/47732476-8c18a380-dc66-11e8-9823-e42979e814ed.png">

turban commented:
Should we add tests for new React components like DataDownloadDialog as well?

turban commented:
![skjermbilde 2018-10-31 kl 16 28 30](https://user-images.githubusercontent.com/548708/47799651-feeb5280-dd2a-11e8-830e-fe5a3cf1ce2e.png)

Some suggestions: 
- A bit more "air" above the select box
- "Please select the format for GeoJSON properties"
- "ID-format" (skip Meta-data...)
- "Use human-readable keys" (less accurate, but I guess most users don't know the meaning og non-dimension attributes).

amcgee commented:
Thanks for the input!

> * Increase spacing above the select field

Done

> * "Please select the format for GeoJSON properties"

I updated it to "Please select the format for dimension attributes", think it's a little more clear

> * "ID-format" (skip Meta-data...)

I changed this to `Meta-data ID sheme` which matches the download option in the Pivot Table and new Data Visualizer download menus.  If we want to update it (and I agree it is not an accessible name), we should do it everywhere.

<img width="144" alt="screenshot 2018-10-31 17 29 56" src="https://user-images.githubusercontent.com/947888/47803393-a9b33f00-dd32-11e8-99ac-21f5deb8925e.png">

> * "Use human-readable keys" (less accurate, but I guess most users don't know the meaning of "non-dimension attributes").

I changes this to `Use human-readable keys for all other attributes` since I added `dimension` above.  Agree it's not the most user-friendly but it's important to differentiate the two in some way, otherwise both inputs seem like they do the same exact thing.  This text may change or go away when we have server-side support for outputIdScheme.

amcgee commented:
Great suggestions!  Updated with those minor changes.

joeDHIS commented:
I'm taking a look at the dialog now - functionality looks good, I may make a few visual changes to make the info clearer. Will post any changes/specs here!

amcgee commented:
Thanks @joeDHIS, two follow-ups:

1) For the title "GeoJSON keys" do you think this is descriptive enough for the user?  I find it ambiguous to have "GeoJSON keys ID format" and "Use human-readable keys" (they sound like the same thing) which is why I called them "dimension attributes" and "all other attributes" explicitly.  Again this may go away soon but worth pointing out.

2) As stated above I like `ID Format` but it's inconsistent with other apps which have a similar feature, where this is called obtusely `Meta-data ID scheme`, should we be clear or consistent here?  I think if we update one we should update all of them.

amcgee commented:
If we want to update the name I will do the same in PRs to the other apps.

amcgee commented:
Made the changes requested by @joeDHIS, dialog screenshots below.  Let's get this merged then address the above questions.

<img width="646" alt="screenshot 2018-11-01 13 02 50" src="https://user-images.githubusercontent.com/947888/47850994-8b534f00-ddd7-11e8-9fce-00c3b7158511.png">
<img width="648" alt="screenshot 2018-11-01 13 03 06" src="https://user-images.githubusercontent.com/947888/47850995-8b534f00-ddd7-11e8-8078-155f3719b19b.png">

amcgee commented:
💯

joeDHIS commented:
> For the title "GeoJSON keys" do you think this is descriptive enough for the user? I find it ambiguous to have "GeoJSON keys ID format" and "Use human-readable keys" (they sound like the same thing) which is why I called them "dimension attributes" and "all other attributes" explicitly. Again this may go away soon but worth pointing out.
I think we need to get some user feedback on this, potentially Scott is the one who could help us. I'm unsure how technically adept the average Maps user is (of course, there is no real average, but we have to land somewhere). 

> As stated above I like ID Format but it's inconsistent with other apps which have a similar feature, where this is called obtusely Meta-data ID scheme, should we be clear or consistent here? I think if we update one we should update all of them.
I think it is best to be clear first and consistent second, otherwise we risk just making the problem even worse by extending the use of weird terms. Let's go forward with ID format and make a point of updating it elsewhere too. (This will be easier said than done, but a good goal.)

Thanks for updating from the specs so fast! 👍
